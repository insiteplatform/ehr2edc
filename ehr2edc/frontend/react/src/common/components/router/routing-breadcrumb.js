import React from "react";
import {Link, Route, Switch} from "react-router-dom";

export function BreadcrumbLink(route) {
    return (
        <Switch>
            {currentBreadcrumbLink(route)}
            {ancestorBreadcrumbLink(route)}
        </Switch>);
}

function currentBreadcrumbLink(route) {
    return <Route path={route.path}
                  exact={true}
                  children={(props) => (
                      <li>
                          {replaceReactRouterParams(route.title, {
                              ...props.match.params,
                              ...route.breadcrumbParams
                          })}
                      </li>)}/>;
}

function ancestorBreadcrumbLink(route) {
    return <Route path={route.path}
                  exact={route.default}
                  children={(props) => {
                      return (<li>
                          <Link to={replaceReactRouterParams(route.path, props.match.params)} {...props}>
                              {replaceReactRouterParams(
                                  route.title,
                                  {
                                      ...props.match.params,
                                      ...route.breadcrumbParams
                                  }
                              )}
                          </Link>
                      </li>)
                  }}/>;
}

function replaceReactRouterParams(toReplace, params) {
    for (let param in params) {
        toReplace = toReplace.split(":" + param).join(params[param])
    }
    return toReplace;
}
