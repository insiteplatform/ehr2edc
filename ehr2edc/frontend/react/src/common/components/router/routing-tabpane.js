import React from "react";
import {Link, Route} from "react-router-dom";

export function TabPaneLink(props) {
    return (<Route path={props.path}
                   exact={true}
                   children={tabPaneLinkWith(props)}/>);
}

function generateActivePanelLink(props) {
    return ({match}) => {
        const path = props.isDefault ? '/' : props.path;
        const className = !!match && match.isExact ? "radioTab active" : "radioTab";
        const linkId = `radioTab${props.title}`;
        return (
            <Link to={path}>
                <input type="radio" name="radioTab" value="true"
                       className={className} id={linkId}/>
                <label htmlFor={linkId}>{props.title}</label>
            </Link>)
    };
}

function generateUnactivePanelLink(props) {
    return ({match}) => {
        const linkId = `radioTab${props.title}`;
        return (
            <div>
            <input type="radio" name="radioTab" className="radioTab disabled" id={linkId}/>
                   <label htmlFor={linkId}>{props.title}</label>
            </div>)
    };
}

function tabPaneLinkWith(props) {
    const {activate} = props;
    return activate ?  generateActivePanelLink(props) : generateUnactivePanelLink(props);
}
