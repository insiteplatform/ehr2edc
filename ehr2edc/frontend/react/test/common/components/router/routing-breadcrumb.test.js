import renderer from "react-test-renderer";
import React from "react";
import {BreadcrumbLink} from "../../../../src/common/components/router/routing-breadcrumb";
import {MemoryRouter} from "react-router-dom";

describe("Breadcrumb correctly render route ", () => {
    test("The component generated current route correctly based on current location", async () =>  {
        const path = '/app/ehr2edc/studies/:studyId/subjects/:subjectId/eventdefinitions/:eventdefinitionId';
        const app = renderer.create(
            <MemoryRouter initialEntries={["/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1"]}>
                <BreadcrumbLink path={path} title='currentRouteTitle'/>
            </MemoryRouter>
            );
        expect(app.toJSON()).toMatchSnapshot();
    });

    test("The component generated ancestor route correctly based on the router", async () =>  {
        const path = '/app/ehr2edc/studies/:studyId/subjects/:subjectId/eventdefinitions/:eventdefinitionId/';

        const app = renderer.create(
            <MemoryRouter initialEntries={["/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/forms"]} keyLength={0}>
                <BreadcrumbLink path={path} title='aAncestorTitle'/>
            </MemoryRouter>
        );

        expect(app.toJSON()).toMatchSnapshot();
    });
});

describe("Breadcrumb correctly render the title", () => {

    let consoleSpy;

    beforeEach(() => {
        if (typeof consoleSpy === "function") {
            consoleSpy.mockRestore();
        }
    });

    test("The component generates the title correctly based on router matches props", async () =>  {
        const path = '/app/ehr2edc/studies/:studyId/subjects/:subjectId/eventdefinitions/:eventdefinitionId';
        const app = mount(
            <MemoryRouter initialEntries={["/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1"]}>
                <BreadcrumbLink path={path} title=':subjectId'/>
            </MemoryRouter>
        );

        expect(app.exists('li')).toBeTruthy();
        expect(app.find('li').text()).toBe("subjectId_1");
    });

    test("The component generates the title correctly based on breadcrumbParameters", async () =>  {
        consoleSpy = jest.spyOn(console, "error").mockImplementation(); // Suppress react-router error
        const path = '/app/ehr2edc/studies/:studyId/subjects/:subjectId/eventdefinitions/:eventdefinitionId';
        const breadcrumbParams = {
            edcSubjectReference: 'edcSubjectReference_1'
        };

        const app = mount(
            <MemoryRouter initialEntries={["/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/form"]}>
                <BreadcrumbLink path={path} title=':edcSubjectReference' breadcrumbParams={breadcrumbParams}/>
            </MemoryRouter>
        );

        expect(app.exists('li')).toBeTruthy();
        expect(app.find('li').text()).toBe("edcSubjectReference_1")
    });
});