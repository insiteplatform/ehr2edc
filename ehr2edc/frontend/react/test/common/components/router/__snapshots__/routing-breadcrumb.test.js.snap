// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb correctly render route  The component generated ancestor route correctly based on the router 1`] = `
<li>
  <a
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/forms",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/forms",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    href="/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/"
    location={
      Object {
        "hash": "",
        "pathname": "/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1/forms",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {
          "eventdefinitionId": "eventDefinitionId_1",
          "studyId": "studyId_1",
          "subjectId": "subjectId_1",
        },
        "path": "/app/ehr2edc/studies/:studyId/subjects/:subjectId/eventdefinitions/:eventdefinitionId/",
        "url": "/app/ehr2edc/studies/studyId_1/subjects/subjectId_1/eventdefinitions/eventDefinitionId_1",
      }
    }
    onClick={[Function]}
  >
    aAncestorTitle
  </a>
</li>
`;

exports[`Breadcrumb correctly render route  The component generated current route correctly based on current location 1`] = `
<li>
  currentRouteTitle
</li>
`;
